packages/web/src/common/store/upload/sagas.ts

```diff
@@ -0,0 +1,1037 @@
+): UploadTrackMetadata => ({
+  ...trackMetadata,
+  preview_start_seconds: trackMetadata.preview_start_seconds ?? null
+})
+
+/**
+ * Combines the metadata for a track and a collection (playlist or album),
+ * taking the metadata from the playlist when the track is missing it.
+ */
+const combineMetadata = (
+  trackMetadata: TrackMetadataForUpload,
+  collectionMetadata: CollectionValues
+) => {
+  const metadata = trackMetadata
+  metadata.artwork = trackMetadata.artwork ?? collectionMetadata.artwork
+
+  if (!metadata.genre)
+    metadata.genre = collectionMetadata.trackDetails.genre ?? ''
+  if (!metadata.mood) metadata.mood = collectionMetadata.trackDetails.mood ?? ''
+  if (!metadata.release_date)
+    metadata.release_date = collectionMetadata.release_date ?? null
+
+  if (metadata.tags === null && collectionMetadata.trackDetails.tags) {
+    // Take collection tags
+    metadata.tags = collectionMetadata.trackDetails.tags
+  }
+  return trackMetadata
+}
+
+/**
+ * Creates a callback that runs on upload progress in libs/sdk and
+ * reports that progress to the store via actions.
+ */
+const makeOnProgress = (
+  trackIndex: number,
+  stemIndex: number | null,
+  progressChannel: Channel<ProgressAction>
+) => {
+  return (progress: Parameters<ProgressCB>[0]) => {
+    const key = 'audio' in progress ? 'audio' : 'art'
+    const p =
+      'audio' in progress
+        ? progress.audio
+        : 'art' in progress
+        ? progress.art
+        : null
+    if (p === null) {
+      return
+    }
+    const { upload, transcode } = p
+    try {
+      progressChannel.put(
+        updateProgress({
+          trackIndex,
+          stemIndex,
+          key,
+          progress: {
+            loaded: upload?.loaded,
+            total: upload?.total,
+            transcode: transcode?.decimal,
+            status:
+              !upload || upload.loaded !== upload.total
+                ? ProgressStatus.UPLOADING
+                : ProgressStatus.PROCESSING
+          }
+        })
+      )
+    } catch {
+      // Sometimes this can fail repeatedly in quick succession (root cause TBD)
+      // it doesn't seem to affect the CX so catch to avoid spamming sentry
+    }
+  }
+}
+
+/**
+ * Deletes a list of tracks by track IDs.
+ * Used in cleaning up orphaned stems or tracks of a collection.
+ */
+function* deleteTracks(trackIds: ID[]) {
+  const audiusBackendInstance = yield* getContext('audiusBackendInstance')
+  const libs = yield* call(audiusBackendInstance.getAudiusLibsTyped)
+
+  yield* all(
+    trackIds.map((id) => call([libs.Track, libs.Track!.deleteTrack], id))
+  )
+}
+
+/**
+ * Calculates the number of workers to process upload requests.
+ * Ensures we don't have absurd amounts of data going over the wire at once.
+ */
+const getNumUploadWorkers = (trackFiles: File[]) => {
+  const largestFileSize = Math.max(...trackFiles.map((t) => t.size))
+
+  // Divide it out so that we never hit > MAX_CONCURRENT_TRACK_SIZE_BYTES in flight.
```

Comment:
haven't taken a look at this code in a while, but... probably can bump this... or maybe even remove it. the browser should just handle that

Reply:
yeah i vote remove